src/transformers/models/ibert/utils/quant_utils.py

```diff
@@ -0,0 +1,264 @@
+            scale = torch.clamp(scale, min=1e-8) / n 
+
+        else:
+            scale = max(saturation_min.abs(), saturation_max.abs())
+            scale = torch.clamp(scale, min=1e-8) / n 
+
+    return scale 
+
+
+class SymmetricQuantFunction(Function):
+    """
+    Class to quantize the given floating-point values using symmetric quantization with given range and bitwidth.
+    """
+    @staticmethod
+    def forward(ctx, x, k, percentile_mode=False, specified_scale=None):
+        """
+        x: floating point tensor to be quantized
+        k: quantization bitwidth
+        Note that the current implementation of SymmetricQuantFunction requires pre-calculated scaling factor.
+        specified_scale: pre-calculated scaling factor for the tensor x
+        """
+        
+        if specified_scale is not None:
+            scale = specified_scale
+
+        zero_point = torch.tensor(0.).cuda()
+
+        n = 2 ** (k - 1) - 1
+        new_quant_x = linear_quantize(x, scale, zero_point, inplace=False)
+        new_quant_x = torch.clamp(new_quant_x, -n, n-1)
+
+        ctx.scale = scale 
+        return new_quant_x
+
+    @staticmethod
+    def backward(ctx, grad_output):
+
+        scale = ctx.scale
+        if len(grad_output.shape) == 4:
+            scale = scale.view(-1, 1, 1, 1)
+        # reshape scale and zeropoint for linear weights
+        elif len(grad_output.shape) == 2:
+            scale = scale.view(-1, 1)
+        else:
+            scale = scale.view(-1)
+
+        return grad_output.clone() / scale, None, None, None, None
+
+
+class floor_ste(Function):
+    """
+    Straight-through Estimator(STE) for torch.floor()
+    """
+    @staticmethod
+    def forward(ctx, x):
+        return torch.floor(x)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.clone()
+
+
+class round_ste(Function):
+    """
+    Straight-through Estimator(STE) for torch.round()
+    """
+    @staticmethod
+    def forward(ctx, x):
+        return torch.round(x)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        return grad_output.clone()
+
+
+def batch_frexp(inputs, max_bit=31):
+    """
+    Decompose the scaling factor into mantissa and twos exponent.
+
+    Parameters:
+    ----------
+    inputs: scaling factor
+    return: (mantissa, exponent)
+    """
+
+    shape_of_input = inputs.size()
+
+    # trans the input to be a 1-d tensor
+    inputs = inputs.view(-1)
+    
+    output_m, output_e = np.frexp(inputs.cpu().numpy())
+    tmp_m = []
+    for m in output_m:
+        int_m_shifted = int(Decimal(m * (2**max_bit)).quantize(Decimal('1'), 
+            rounding=decimal.ROUND_HALF_UP))
+        tmp_m.append(int_m_shifted)
+    output_m = np.array(tmp_m)
+
+    output_e = float(max_bit) - output_e
+
+    return torch.from_numpy( output_m ).cuda().view(shape_of_input), \
+           torch.from_numpy( output_e ).cuda().view(shape_of_input)
+
+class fixedpoint_mul(Function):
```

Comment:
even though this class is a `Function`, I think it's better to call it `FixedPointMul`